File: runtime/go/cable.go
// Roundhouse Go cable runtime.
//
// Action Cable WebSocket + Turbo Streams broadcaster. Mirrors
// runtime/rust/cable.rs + runtime/python/cable.py — same wire
// format (actioncable-v1-json), same partial-renderer registry,
// same per-channel subscriber map.
//
// Uses nhooyr.io/websocket — added to go.mod by the emitter and
// resolved via `go mod tidy` on first build.

package app

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"net/http"
	"strings"
	"sync"
	"time"

	"nhooyr.io/websocket"
)

// ── Partial-renderer registry ─────────────────────────────────

var (
	partialMu        sync.RWMutex
	partialRenderers = map[string]func(int64) string{}
)

// RegisterPartial associates a model class name with a renderer
// that takes a record id and returns the partial HTML. Cable's
// broadcast_*_to helpers consult this registry to produce frames
// without needing model-specific imports at the cable-runtime
// layer.
func RegisterPartial(typeName string, fn func(int64) string) {
	partialMu.Lock()
	defer partialMu.Unlock()
	partialRenderers[typeName] = fn
}

// RenderPartial looks up and invokes a registered renderer.
// Returns a placeholder div when no renderer is registered so
// callers never panic on an unknown type.
func RenderPartial(typeName string, id int64) string {
	partialMu.RLock()
	fn := partialRenderers[typeName]
	partialMu.RUnlock()
	if fn == nil {
		return "<div>" + typeName + " #" + itoa64(id) + "</div>"
	}
	return fn(id)
}

// ── Turbo Streams rendering ───────────────────────────────────

// TurboStreamHTML wraps a rendered partial in the Turbo Stream
// element Turbo's client splices into the DOM. Empty content
// collapses to a self-closing template (used by `remove`).
func TurboStreamHTML(action, target, content string) string {
	if content == "" {
		return `<turbo-stream action="` + action + `" target="` + target + `"></turbo-stream>`
	}
	return `<turbo-stream action="` + action + `" target="` + target + `"><template>` + content + `</template></turbo-stream>`
}

func domIDFor(table string, id int64) string {
	singular := table
	if strings.HasSuffix(singular, "s") {
		singular = singular[:len(singular)-1]
	}
	return singular + "_" + itoa64(id)
}

// ── Broadcast helpers ─────────────────────────────────────────

// BroadcastReplaceTo replaces the target element with the record's
// partial. Empty target defaults to `<singular>_<id>`.
func BroadcastReplaceTo(table string, id int64, typeName, channel, target string) {
	t := target
	if t == "" {
		t = domIDFor(table, id)
	}
	html := RenderPartial(typeName, id)
	dispatch(channel, TurboStreamHTML("replace", t, html))
}

// BroadcastPrependTo prepends the partial into the target
// container. Empty target defaults to the table name (the scaffold
// convention: `<ul id="articles">`).
func BroadcastPrependTo(table string, id int64, typeName, channel, target string) {
	t := target
	if t == "" {
		t = table
	}
	html := RenderPartial(typeName, id)
	dispatch(channel, TurboStreamHTML("prepend", t, html))
}

// BroadcastAppendTo appends the partial into the target container.
func BroadcastAppendTo(table string, id int64, typeName, channel, target string) {
	t := target
	if t == "" {
		t = table
	}
	html := RenderPartial(typeName, id)
	dispatch(channel, TurboStreamHTML("append", t, html))
}

// BroadcastRemoveTo removes the target element. Empty target
// defaults to `<singular>_<id>`.
func BroadcastRemoveTo(table string, id int64, channel, target string) {
	t := target
	if t == "" {
		t = domIDFor(table, id)
	}
	dispatch(channel, TurboStreamHTML("remove", t, ""))
}

// ── Subscriber registry + dispatch ────────────────────────────

type cableSub struct {
	conn       *websocket.Conn
	identifier string
}

var (
	subsMu      sync.Mutex
	subscribers = map[string][]*cableSub{}
)

func dispatch(channel, html string) {
	subsMu.Lock()
	subs := append([]*cableSub(nil), subscribers[channel]...)
	subsMu.Unlock()
	for _, s := range subs {
		payload, err := json.Marshal(map[string]any{
			"type":       "message",
			"identifier": s.identifier,
			"message":    html,
		})
		if err != nil {
			continue
		}
		// Best-effort write with a short timeout so a stalled
		// client doesn't block the broadcaster.
		ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
		_ = s.conn.Write(ctx, websocket.MessageText, payload)
		cancel()
	}
}

// Broadcast is the stub kept for back-compat with the earlier
// server wiring — the `/cable` handler now upgrades.
func Broadcast(channel, html string) { dispatch(channel, html) }

// ── WebSocket handler ─────────────────────────────────────────

// CableHandler is mounted at `/cable` by the server. It
// negotiates the actioncable-v1-json subprotocol, sends welcome +
// ping frames, handles subscribe commands, and cleans up
// subscriptions on close.
func CableHandler(w http.ResponseWriter, r *http.Request) {
	conn, err := websocket.Accept(w, r, &websocket.AcceptOptions{
		Subprotocols: []string{"actioncable-v1-json"},
	})
	if err != nil {
		return
	}
	defer conn.Close(websocket.StatusInternalError, "cable handler exiting")

	ctx, cancel := context.WithCancel(r.Context())
	defer cancel()

	// Welcome frame.
	welcome, _ := json.Marshal(map[string]string{"type": "welcome"})
	if err := conn.Write(ctx, websocket.MessageText, welcome); err != nil {
		return
	}

	// Ping loop.
	go func() {
		ticker := time.NewTicker(3 * time.Second)
		defer ticker.Stop()
		for {
			select {
			case <-ctx.Done():
				return
			case <-ticker.C:
				payload, _ := json.Marshal(map[string]any{
					"type":    "ping",
					"message": time.Now().Unix(),
				})
				wctx, wcancel := context.WithTimeout(ctx, 2*time.Second)
				err := conn.Write(wctx, websocket.MessageText, payload)
				wcancel()
				if err != nil {
					return
				}
			}
		}
	}()

	var (
		myMu  sync.Mutex
		myEnt []struct {
			channel string
			sub     *cableSub
		}
	)

	// Subscribe handler loop.
	for {
		_, data, err := conn.Read(ctx)
		if err != nil {
			break
		}
		var payload map[string]any
		if err := json.Unmarshal(data, &payload); err != nil {
			continue
		}
		cmd, _ := payload["command"].(string)
		if cmd != "subscribe" {
			continue
		}
		identifier, _ := payload["identifier"].(string)
		if identifier == "" {
			continue
		}
		channel := decodeChannel(identifier)
		if channel == "" {
			continue
		}
		sub := &cableSub{conn: conn, identifier: identifier}
		subsMu.Lock()
		subscribers[channel] = append(subscribers[channel], sub)
		subsMu.Unlock()
		myMu.Lock()
		myEnt = append(myEnt, struct {
			channel string
			sub     *cableSub
		}{channel, sub})
		myMu.Unlock()
		confirm, _ := json.Marshal(map[string]string{
			"type":       "confirm_subscription",
			"identifier": identifier,
		})
		_ = conn.Write(ctx, websocket.MessageText, confirm)
	}

	// Cleanup on disconnect.
	myMu.Lock()
	ents := myEnt
	myMu.Unlock()
	subsMu.Lock()
	for _, e := range ents {
		list := subscribers[e.channel]
		for i, s := range list {
			if s == e.sub {
				subscribers[e.channel] = append(list[:i], list[i+1:]...)
				break
			}
		}
		if len(subscribers[e.channel]) == 0 {
			delete(subscribers, e.channel)
		}
	}
	subsMu.Unlock()
	conn.Close(websocket.StatusNormalClosure, "")
}

// decodeChannel recovers the channel name from Turbo's signed
// stream identifier. The identifier is a JSON blob like
// `{"channel":"Turbo::StreamsChannel",
//
//	"signed_stream_name":"<base64>--<digest>"}`;
//
// the base64 prefix holds a JSON-encoded channel name. Returns ""
// on malformed input so the handler silently drops the frame.
func decodeChannel(identifier string) string {
	var id map[string]any
	if err := json.Unmarshal([]byte(identifier), &id); err != nil {
		return ""
	}
	signed, ok := id["signed_stream_name"].(string)
	if !ok {
		return ""
	}
	b64 := signed
	if idx := strings.Index(signed, "--"); idx >= 0 {
		b64 = signed[:idx]
	}
	decoded, err := base64.StdEncoding.DecodeString(b64)
	if err != nil {
		return ""
	}
	var channel string
	if err := json.Unmarshal(decoded, &channel); err != nil {
		return ""
	}
	return channel
}

// itoa64 is a no-alloc local int64→string since strconv.FormatInt
// is in strconv which the cable runtime doesn't otherwise need.
func itoa64(n int64) string {
	if n == 0 {
		return "0"
	}
	var buf [20]byte
	i := len(buf)
	neg := n < 0
	if neg {
		n = -n
	}
	for n > 0 {
		i--
		buf[i] = byte('0' + n%10)
		n /= 10
	}
	if neg {
		i--
		buf[i] = '-'
	}
	return string(buf[i:])
}

File: runtime/rust/cable.rs
//! Action Cable server + Turbo Streams broadcaster.
//!
//! Mirrors the TS runtime's `CableServer` + `/cable` handler.
//! Ported from railcar's proven implementation (pings + the
//! `actioncable-v1-json` subprotocol are both known-good there).
//!
//! Two halves:
//!   1. Broadcasting — models call `broadcast_prepend_to` /
//!      `broadcast_replace_to` / `broadcast_remove_to`, which render
//!      the appropriate `<turbo-stream>` element and push it to
//!      every subscriber of the given channel. A partial-renderer
//!      registry (`register_partial`) lets the runtime reach back
//!      into the generated `views::*` functions without this file
//!      having to know model-specific types.
//!   2. WebSocket — `cable_handler` upgrades incoming requests,
//!      sends a welcome frame, pings every 3s, and on `subscribe`
//!      commands decodes Turbo's signed-stream-name blob to recover
//!      the channel name, then registers the socket's outbound mpsc
//!      sender with the global `CABLE` registry.
//!
//! Generated code (Phase B of the cable work) calls these via
//! `crate::cable::...` from `impl Broadcaster for <Model>` blocks
//! produced by the emitter's `broadcasts_to` translation.
//!
//! The `actioncable-v1-json` subprotocol spec:
//!   https://github.com/rails/rails/blob/main/actioncable/lib/action_cable/server/worker.rb
//! Frames used here: `welcome`, `ping`, `confirm_subscription`,
//! `message`. Rejections + `unsubscribe` commands aren't needed
//! for the current broadcast paths.

use std::collections::HashMap;
use std::sync::{LazyLock, RwLock};
use std::time::{SystemTime, UNIX_EPOCH};

use axum::extract::ws::{Message, WebSocket, WebSocketUpgrade};
use axum::response::IntoResponse;
use base64::Engine;
use futures_util::{SinkExt, StreamExt};
use serde_json::{json, Value};
use tokio::sync::mpsc;

// ── Partial-renderer registry ──────────────────────────────────
//
// Models register a closure that renders an instance identified by
// id into its Turbo Stream partial HTML. Kept as a runtime lookup
// (rather than parameterising Broadcaster on the model type) so
// broadcasts called on associations — e.g., `comment.article`'s
// replace broadcast — can find the parent's partial without the
// child model needing to know the parent's view module.

pub type RenderPartialFn = Box<dyn Fn(i64) -> String + Send + Sync>;

static PARTIAL_RENDERERS: LazyLock<RwLock<HashMap<String, RenderPartialFn>>> =
    LazyLock::new(|| RwLock::new(HashMap::new()));

/// Register a partial renderer for `type_name` (the model class
/// name, e.g. `"Article"`). The closure receives a record id and
/// returns the rendered partial HTML, or empty string on miss.
pub fn register_partial(type_name: &str, f: impl Fn(i64) -> String + Send + Sync + 'static) {
    PARTIAL_RENDERERS
        .write()
        .expect("cable partial renderers poisoned")
        .insert(type_name.to_string(), Box::new(f));
}

/// Look up and invoke a registered partial renderer. Returns a
/// placeholder div when no renderer is registered — tests can
/// assert on the fallback rather than panicking.
pub fn render_partial(type_name: &str, id: i64) -> String {
    let table = PARTIAL_RENDERERS
        .read()
        .expect("cable partial renderers poisoned");
    match table.get(type_name) {
        Some(f) => f(id),
        None => format!("<div>{} #{}</div>", type_name, id),
    }
}

// ── Turbo Streams rendering ────────────────────────────────────

/// Render a single `<turbo-stream>` element. Empty content collapses
/// to a self-closing template (used by `remove` actions).
pub fn turbo_stream_html(action: &str, target: &str, content: &str) -> String {
    if content.is_empty() {
        format!(
            r#"<turbo-stream action="{}" target="{}"></turbo-stream>"#,
            action, target
        )
    } else {
        format!(
            r#"<turbo-stream action="{}" target="{}"><template>{}</template></turbo-stream>"#,
            action, target, content,
        )
    }
}

/// Rails convention: `<singular>_<id>`. Naive depluralise — strips
/// a trailing `s` if present. Matches railcar's `dom_id_for`.
fn dom_id_for(table_name: &str, id: i64) -> String {
    let singular = table_name.strip_suffix('s').unwrap_or(table_name);
    format!("{}_{}", singular, id)
}

// ── Broadcast helpers ──────────────────────────────────────────

/// Replace the target element with the record's partial. Defaults
/// `target` to `<singular>_<id>` when caller passes an empty
/// string (matches Rails' `broadcast_replace_to` with no explicit
/// target).
pub fn broadcast_replace_to(
    table_name: &str,
    id: i64,
    type_name: &str,
    channel: &str,
    target: &str,
) {
    let target = if target.is_empty() {
        dom_id_for(table_name, id)
    } else {
        target.to_string()
    };
    let html = render_partial(type_name, id);
    let stream = turbo_stream_html("replace", &target, &html);
    CABLE.broadcast(channel, &stream);
}

/// Prepend the record's partial into the target container.
/// Defaults `target` to the table name (the scaffold's `<ul
/// id="articles">` convention).
pub fn broadcast_prepend_to(
    table_name: &str,
    id: i64,
    type_name: &str,
    channel: &str,
    target: &str,
) {
    let target = if target.is_empty() {
        table_name.to_string()
    } else {
        target.to_string()
    };
    let html = render_partial(type_name, id);
    let stream = turbo_stream_html("prepend", &target, &html);
    CABLE.broadcast(channel, &stream);
}

/// Append the record's partial into the target container. Same
/// default-target rule as prepend.
pub fn broadcast_append_to(
    table_name: &str,
    id: i64,
    type_name: &str,
    channel: &str,
    target: &str,
) {
    let target = if target.is_empty() {
        table_name.to_string()
    } else {
        target.to_string()
    };
    let html = render_partial(type_name, id);
    let stream = turbo_stream_html("append", &target, &html);
    CABLE.broadcast(channel, &stream);
}

/// Remove the target element. Target defaults to `<singular>_<id>`
/// so `broadcast_remove_to(channel)` on a record deletes its own
/// DOM node.
pub fn broadcast_remove_to(table_name: &str, id: i64, channel: &str, target: &str) {
    let target = if target.is_empty() {
        dom_id_for(table_name, id)
    } else {
        target.to_string()
    };
    let stream = turbo_stream_html("remove", &target, "");
    CABLE.broadcast(channel, &stream);
}

// ── CableServer ────────────────────────────────────────────────

struct Subscriber {
    /// Outbound channel to this socket's send task. Cloned into the
    /// registry; the socket task owns the receiver half.
    tx: mpsc::UnboundedSender<String>,
    /// The raw identifier string the client sent on subscribe. We
    /// echo it back in every broadcast so Turbo can route the
    /// message to the right `<turbo-cable-stream-source>` element.
    identifier: String,
}

pub struct CableServer {
    channels: RwLock<HashMap<String, Vec<Subscriber>>>,
}

/// Process-wide registry. One per server; fine as a static because
/// the server runs one app per process (same assumption as
/// `server::LAYOUT_FN`).
pub static CABLE: LazyLock<CableServer> = LazyLock::new(|| CableServer {
    channels: RwLock::new(HashMap::new()),
});

impl CableServer {
    fn subscribe(&self, channel: &str, tx: mpsc::UnboundedSender<String>, identifier: &str) {
        self.channels
            .write()
            .expect("cable channels poisoned")
            .entry(channel.to_string())
            .or_default()
            .push(Subscriber {
                tx,
                identifier: identifier.to_string(),
            });
    }

    /// Drop any subscribers whose `tx` matches the given pointer.
    /// Called on socket close. We compare by pointer rather than
    /// PartialEq because `UnboundedSender` doesn't implement it;
    /// each subscriber holds a distinct sender, so pointer identity
    /// is sufficient and avoids threading a subscriber id through
    /// the WebSocket task.
    fn unsubscribe(&self, tx_ptr: usize) {
        let mut channels = self.channels.write().expect("cable channels poisoned");
        for subs in channels.values_mut() {
            subs.retain(|s| &s.tx as *const _ as usize != tx_ptr);
        }
        channels.retain(|_, subs| !subs.is_empty());
    }

    /// Push `html` as a Turbo Stream `message` frame to every
    /// subscriber on `channel`. Dropped senders are ignored — the
    /// subscriber will be cleaned up on the close-driven
    /// `unsubscribe` path.
    pub fn broadcast(&self, channel: &str, html: &str) {
        let channels = self.channels.read().expect("cable channels poisoned");
        if let Some(subs) = channels.get(channel) {
            for sub in subs {
                let frame = json!({
                    "type": "message",
                    "identifier": sub.identifier,
                    "message": html,
                })
                .to_string();
                let _ = sub.tx.send(frame);
            }
        }
    }
}

// ── WebSocket handler ──────────────────────────────────────────

/// Axum handler for `GET /cable`. Negotiates the
/// `actioncable-v1-json` subprotocol (Turbo's client requires the
/// echo) and hands off to the per-socket task.
pub async fn cable_handler(ws: WebSocketUpgrade) -> impl IntoResponse {
    ws.protocols(["actioncable-v1-json"])
        .on_upgrade(handle_socket)
}

async fn handle_socket(socket: WebSocket) {
    let (mut sender, mut receiver) = socket.split();

    // Welcome frame — the Action Cable client waits for this before
    // it sends its first `subscribe`.
    if sender
        .send(Message::Text(
            json!({"type": "welcome"}).to_string().into(),
        ))
        .await
        .is_err()
    {
        return;
    }

    // Single outbound channel merges broadcasts + pings onto the
    // shared sender half. Cloning the tx into the ping task and
    // the registry lets each source push independently without
    // locking the socket writer.
    let (tx, mut rx) = mpsc::unbounded_channel::<String>();
    let tx_ptr = &tx as *const _ as usize;

    let ping_tx = tx.clone();
    let ping_task = tokio::spawn(async move {
        let mut interval = tokio::time::interval(std::time::Duration::from_secs(3));
        // First tick fires immediately; skip it so we don't ping
        // before the welcome + confirm_subscription round-trip.
        interval.tick().await;
        loop {
            interval.tick().await;
            let ts = SystemTime::now()
                .duration_since(UNIX_EPOCH)
                .map(|d| d.as_secs() as i64)
                .unwrap_or(0);
            let frame = json!({"type": "ping", "message": ts}).to_string();
            if ping_tx.send(frame).is_err() {
                break;
            }
        }
    });

    let send_task = tokio::spawn(async move {
        while let Some(frame) = rx.recv().await {
            if sender.send(Message::Text(frame.into())).await.is_err() {
                break;
            }
        }
    });

    while let Some(Ok(msg)) = receiver.next().await {
        let Message::Text(text) = msg else { continue };
        let Ok(payload) = serde_json::from_str::<Value>(&text) else {
            continue;
        };
        if payload.get("command").and_then(Value::as_str) != Some("subscribe") {
            continue;
        }
        let Some(identifier) = payload.get("identifier").and_then(Value::as_str) else {
            continue;
        };
        let Some(channel) = decode_channel(identifier) else {
            continue;
        };
        CABLE.subscribe(&channel, tx.clone(), identifier);
        let confirm = json!({
            "type": "confirm_subscription",
            "identifier": identifier,
        })
        .to_string();
        let _ = tx.send(confirm);
    }

    ping_task.abort();
    send_task.abort();
    CABLE.unsubscribe(tx_ptr);
}

/// Recover the channel name from Turbo's `signed_stream_name`.
/// The identifier is a JSON blob like
/// `{"channel":"Turbo::StreamsChannel","signed_stream_name":"<base64>--<digest>"}`;
/// the base64 segment holds a JSON-encoded channel name (e.g.
/// `"articles"`). If either decode fails we fall back to the raw
/// identifier so tests can subscribe by literal channel string.
fn decode_channel(identifier: &str) -> Option<String> {
    let id_json = serde_json::from_str::<Value>(identifier).ok()?;
    let signed = id_json
        .get("signed_stream_name")
        .and_then(Value::as_str)?;
    let base64_part = signed.split("--").next().unwrap_or("");
    let decoded_bytes = base64::engine::general_purpose::STANDARD
        .decode(base64_part)
        .ok()?;
    let decoded = std::str::from_utf8(&decoded_bytes).ok()?;
    serde_json::from_str::<String>(decoded).ok()
}

// ── Broadcaster trait ──────────────────────────────────────────

/// Implemented on models with `broadcasts_to` declarations. The
/// emitter's `broadcasts_to` translation (Phase B) generates these
/// implementations; the runtime calls them from the generated
/// `save()` / `destroy()` methods at the end of a successful
/// persist.
///
/// Kept separate from `Model` so that models without any broadcast
/// hooks don't need a stub impl — the emitter only emits this for
/// models that declare `broadcasts_to`, and the save/destroy
/// codegen conditionally calls into it.
pub trait Broadcaster {
    fn after_save(&self);
    fn after_delete(&self);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn turbo_stream_html_wraps_content_in_template() {
        let got = turbo_stream_html("replace", "article_1", "<div>hi</div>");
        assert_eq!(
            got,
            r#"<turbo-stream action="replace" target="article_1"><template><div>hi</div></template></turbo-stream>"#
        );
    }

    #[test]
    fn turbo_stream_html_self_closes_when_empty() {
        let got = turbo_stream_html("remove", "article_1", "");
        assert_eq!(
            got,
            r#"<turbo-stream action="remove" target="article_1"></turbo-stream>"#
        );
    }

    #[test]
    fn dom_id_for_strips_trailing_s() {
        assert_eq!(dom_id_for("articles", 7), "article_7");
        assert_eq!(dom_id_for("comment", 3), "comment_3");
    }

    #[test]
    fn render_partial_falls_back_when_unregistered() {
        // Use a distinct type name so parallel tests don't collide.
        let got = render_partial("UnregisteredNoise", 99);
        assert_eq!(got, "<div>UnregisteredNoise #99</div>");
    }

    #[test]
    fn decode_channel_recovers_plain_base64_name() {
        // Construct a signed stream name for `"articles"`.
        let inner = serde_json::to_string("articles").unwrap();
        let b64 =
            base64::engine::general_purpose::STANDARD.encode(inner.as_bytes());
        let signed = format!("{}--unsigned", b64);
        let identifier = serde_json::json!({
            "channel": "Turbo::StreamsChannel",
            "signed_stream_name": signed,
        })
        .to_string();
        assert_eq!(decode_channel(&identifier).as_deref(), Some("articles"));
    }

    #[test]
    fn decode_channel_returns_none_on_bad_input() {
        assert!(decode_channel("not json").is_none());
        assert!(decode_channel(r#"{"no":"signed"}"#).is_none());
    }
}
